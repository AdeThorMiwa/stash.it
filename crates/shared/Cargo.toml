[package]
name = "shared"
version = "0.1.0"
edition = "2024"

[dependencies]
uuid = { version = "1.18.1", features = ["v4"] }
thiserror = { workspace = true }
serde = { workspace = true }
chrono = { workspace = true }
async-trait = { workspace = true }
alloy = "1.0.41"
more-di = { workspace = true }
tokio = { workspace = true }
config = { workspace = true }

[features]
testing = []
