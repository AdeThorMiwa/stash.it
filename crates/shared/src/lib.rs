pub mod domain;
pub mod infrastructure;
#[cfg(feature = "testing")]
pub mod testing;
