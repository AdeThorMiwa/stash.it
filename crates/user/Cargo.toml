[package]
name = "user"
version = "0.1.0"
edition = "2024"

[dependencies]
shared = { path = "../shared" }
macros = { path = "../macros" }
email_address = "0.2.9"
serde = { workspace = true }
chrono = { workspace = true }
async-trait = { workspace = true }
more-di = { workspace = true }
once_cell = { workspace = true }
send_wrapper = { workspace = true }
ctor = { workspace = true }
rand = "0.9.2"
thiserror = { workspace = true }
jsonwebtoken = { version = "10.0.0", features = ["aws_lc_rs"] }
derive_builder = { workspace = true }
config = { workspace = true }
tokio = { workspace = true }
regex = { workspace = true }
