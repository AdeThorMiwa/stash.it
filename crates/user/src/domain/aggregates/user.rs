use crate::domain::value_objects::{email::EmailAddress, user_status::UserStatus};
use chrono::Utc;
use shared::domain::value_objects::{date::Date, pid::Pid};

#[derive(Debug, Clone)]
pub struct User {
    pub(crate) pid: Pid,
    pub(crate) email: EmailAddress,
    pub(crate) status: UserStatus,
    pub(crate) created_at: Date,
    pub(crate) last_login_at: Date,
}

impl User {
    /// create a new user
    pub fn new(email: EmailAddress) -> Self {
        let pid = Pid::new();

        Self {
            pid,
            email,
            status: UserStatus::PendingProfile,
            created_at: Utc::now(),
            last_login_at: Utc::now(),
        }
    }

    /// update user status to `new_status`
    pub fn update_status(&mut self, new_status: UserStatus) {
        self.status = new_status;
    }

    /// update user last login time to now
    pub fn update_last_login(&mut self) {
        self.last_login_at = Utc::now()
    }
}
