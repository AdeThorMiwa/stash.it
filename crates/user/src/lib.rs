use macros::initialize_container;

pub mod application;
mod domain;
mod infrastructure;

initialize_container!();

