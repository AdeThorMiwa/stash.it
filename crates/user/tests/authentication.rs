use crate::common::{bootstrap::bootstrap, insta_filters::redactions::cleanup_model_generics, string_utils::extract_otp};
use insta::{assert_debug_snapshot, with_settings};
use shared::{
    domain::value_objects::pid::Pid,
    infrastructure::{mailing::Mailer, types::Result},
};
use std::{str::FromStr, sync::Arc};
use user::{
    application::{authentication::AuthenticationService, user_management::UserManagementService},
    domain::value_objects::email::EmailAddress,
    infrastructure::auth::jwt_service::JWTService,
};

mod common;

#[tokio::test]
async fn can_request_authentication_code() -> Result<()> {
    // Arrange
    configure_insta!();
    let provider = bootstrap();
    let mailer: Arc<dyn Mailer> = provider.get_required();
    let authentication_service: Arc<AuthenticationService> = provider.get_required();
    let email = EmailAddress::from_str("tom@stash.it").unwrap();

    // Act
    let result = authentication_service.request_authentication_code(&email).await;
    let deliveries = mailer.deliveries().await;
    let code = extract_otp(&deliveries.messages.first().unwrap()).unwrap();

    // Assert
    assert!(result.is_ok(), "Request authentication code should succeed");
    assert_eq!(deliveries.count, 1, "Mailer should have delivered 1 email");
    assert_eq!(&code, "112358");
    Ok(())
}

#[tokio::test]
async fn can_authenticate() -> Result<()> {
    // Arrange
    configure_insta!();
    let provider = bootstrap();
    let mailer: Arc<dyn Mailer> = provider.get_required();
    let authentication_service: Arc<AuthenticationService> = provider.get_required();
    let email = EmailAddress::from_str("test@stash.it").unwrap();

    // Act
    let session_id = authentication_service.request_authentication_code(&email).await?;
    let deliveries = mailer.deliveries().await;
    let code = extract_otp(&deliveries.messages.first().unwrap()).unwrap();
    let token = authentication_service.authenticate(&session_id, &code).await?;

    let jwt_service: Arc<JWTService> = provider.get_required();
    let claims = jwt_service.decode_token(&token)?;
    let pid = Pid::from_str(claims.sub.as_str()).unwrap();

    let user_service: Arc<UserManagementService> = provider.get_required();
    let user = user_service.get_user_by_pid(&pid).await?.unwrap();

    with_settings!({
        filters => cleanup_model_generics(),
        snapshot_suffix => "authenticated_user"
    }, {
        assert_debug_snapshot!(user);
    });

    Ok(())
}
