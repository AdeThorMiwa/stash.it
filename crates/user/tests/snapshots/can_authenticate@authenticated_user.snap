---
source: crates/user/tests/authentication.rs
expression: user
---
User {
    pid: Pid(
        PID,
    ),
    email: EmailAddress(
        EmailAddress(
            "test@stash.it",
        ),
    ),
    status: PendingProfile,
    created_at: DATEZ,
    last_login_at: DATEZ,
}
